server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -11,32 +11,54 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.apache.lucene.codecs.CodecUtil;
+import org.apache.lucene.index.IndexFileNames;
+import org.apache.lucene.index.SegmentInfos;
 import org.apache.lucene.search.ReferenceManager;
 import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.FilterDirectory;
 import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.opensearch.common.concurrent.GatedCloseable;
+import org.opensearch.index.engine.EngineException;
+import org.opensearch.index.store.RemoteSegmentStoreDirectory;
 
 import java.io.IOException;
-import java.nio.file.NoSuchFileException;
-import java.util.Arrays;
-import java.util.HashSet;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
 
 /**
  * RefreshListener implementation to upload newly created segment files to the remote store
+ *
+ * @opensearch.internal
  */
-public class RemoteStoreRefreshListener implements ReferenceManager.RefreshListener {
+public final class RemoteStoreRefreshListener implements ReferenceManager.RefreshListener {
+    // Visible for testing
+    static final Set<String> EXCLUDE_FILES = Set.of("write.lock");
+    // Visible for testing
+    static final int LAST_N_METADATA_FILES_TO_KEEP = 10;
 
+    private final IndexShard indexShard;
     private final Directory storeDirectory;
-    private final Directory remoteDirectory;
-    // ToDo: This can be a map with metadata of the uploaded file as value of the map (GitHub #3398)
-    private final Set<String> filesUploadedToRemoteStore;
+    private final RemoteSegmentStoreDirectory remoteDirectory;
+    private final Map<String, String> localSegmentChecksumMap;
+    private long primaryTerm;
     private static final Logger logger = LogManager.getLogger(RemoteStoreRefreshListener.class);
 
-    public RemoteStoreRefreshListener(Directory storeDirectory, Directory remoteDirectory) throws IOException {
-        this.storeDirectory = storeDirectory;
-        this.remoteDirectory = remoteDirectory;
-        // ToDo: Handle failures in reading list of files (GitHub #3397)
-        this.filesUploadedToRemoteStore = new HashSet<>(Arrays.asList(remoteDirectory.listAll()));
+    public RemoteStoreRefreshListener(IndexShard indexShard) {
+        this.indexShard = indexShard;
+        this.storeDirectory = indexShard.store().directory();
+        this.remoteDirectory = (RemoteSegmentStoreDirectory) ((FilterDirectory) ((FilterDirectory) indexShard.remoteStore().directory())
+            .getDelegate()).getDelegate();
+        this.primaryTerm = indexShard.getOperationPrimaryTerm();
+        localSegmentChecksumMap = new HashMap<>();
```

Comment:
would have preferred individual dependencies as supplier rather than a full `IndexShard` dependency, to protect against misuse

Reply:
Yeah, I thought about that but there are different IndexShard methods (`getOperationPrimaryTerm`, `getSegmentInfosSnapshot`) that are getting called in the `afterRefresh()` flow. 

Created https://github.com/opensearch-project/OpenSearch/issues/4316 to track this.